internal/yggdrasil/yggdrasil_test.go

```diff
@@ -586,6 +592,643 @@ var _ = Describe("Yggdrasil", func() {
+
+			// then
+			Expect(res).To(Equal(operations.NewGetDataMessageForDeviceInternalServerError()))
+		})
+
+		It("Secrets missing secret", func() {
+			// given
+			deviceName := "foo"
+			device := getDevice(deviceName)
+			device.Status.Deployments = []v1alpha1.Deployment{{Name: "workload1"}}
+
+			edgeDeviceRepoMock.EXPECT().
+				Read(gomock.Any(), deviceName, testNamespace).
+				Return(device, nil).
+				Times(1)
+
+			secretName := "test"
+			secretNamespacedName := types.NamespacedName{Namespace: device.Namespace, Name: secretName}
+			podData := v1alpha1.Pod{
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:  "test",
+							Image: "test",
+							EnvFrom: []corev1.EnvFromSource{
+								{
+									SecretRef: &corev1.SecretEnvSource{
+										LocalObjectReference: corev1.LocalObjectReference{
+											Name: secretName,
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			}
+
+			deploymentData := &v1alpha1.EdgeDeployment{
+				ObjectMeta: v1.ObjectMeta{
+					Name:      "workload1",
+					Namespace: "default",
+				},
+				Spec: v1alpha1.EdgeDeploymentSpec{
+					DeviceSelector: &v1.LabelSelector{
+						MatchLabels: map[string]string{"test": "test"},
+					},
+					Type: "pod",
+					Pod:  podData,
+					Data: &v1alpha1.DataConfiguration{},
+				}}
+			deployRepoMock.EXPECT().
+				Read(gomock.Any(), "workload1", testNamespace).
+				Return(deploymentData, nil)
+			k8sClient.EXPECT().
+				Get(gomock.Any(), secretNamespacedName, gomock.Any()).
+				Return(errorNotFound)
+
+			// when
+			res := handler.GetDataMessageForDevice(context.TODO(), params)
+
+			// then
+			Expect(res).To(Equal(operations.NewGetDataMessageForDeviceInternalServerError()))
+		})
+
+		It("Secrets partially optional secret", func() {
+			// given
+			deviceName := "foo"
+			device := getDevice(deviceName)
+			device.Status.Deployments = []v1alpha1.Deployment{{Name: "workload1"}}
+
+			edgeDeviceRepoMock.EXPECT().
+				Read(gomock.Any(), deviceName, testNamespace).
+				Return(device, nil).
+				Times(1)
+
+			booltrue := true
+			secretName := "test"
+			secretNamespacedName := types.NamespacedName{Namespace: device.Namespace, Name: secretName}
+			podData := v1alpha1.Pod{
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:  "test",
+							Image: "test",
+							EnvFrom: []corev1.EnvFromSource{
+								{
+									SecretRef: &corev1.SecretEnvSource{
+										LocalObjectReference: corev1.LocalObjectReference{
+											Name: secretName,
+										},
+									},
+								},
+							},
+							Env: []corev1.EnvVar{
+								{
+									Name: "test",
+									ValueFrom: &corev1.EnvVarSource{
+										SecretKeyRef: &corev1.SecretKeySelector{
+											LocalObjectReference: corev1.LocalObjectReference{
+												Name: secretName,
+											},
+											Key:      "key",
+											Optional: &booltrue,
```

Comment:
why you defined boolTrue and not using true here?

Reply:
It requires a pointer. Can use constant as pointer

Reply:
Ok, then defined on the Context of the test, no?

Reply:
We cna define it anywhere. It doesn't matter where and whether once or multiple times.